python/ray/util/collective/collective.py

```diff
@@ -0,0 +1,261 @@
+                ray.kill(store)
+        # Release the communicator resources
+        g.destroy_group()
+
+
+_group_mgr = GroupManager()
+
+
+def is_group_initialized(group_name):
+    """Check if the group is initialized in this process by the group name."""
+    if not _group_mgr.is_group_exist(group_name):
+        return False
+    return True
+
+
+def init_collective_group(world_size: int,
+                          rank: int,
+                          backend=types.Backend.NCCL,
+                          group_name: str = "default"):
+    """
+    Initialize a collective group inside an actor process.
+
+    This is an
+    Args:
+        world_size:
+        rank:
+        backend:
+        group_name:
+
+    Returns:
+        None
+    """
+    backend = types.Backend(backend)
+    _check_backend_availability(backend)
+    global _group_mgr
+    # TODO(Hao): implement a group auto-counter.
+    if not group_name:
+        raise ValueError("group_name '{}' needs to be a string."
+                         .format(group_name))
+
+    if _group_mgr.is_group_exist(group_name):
+        raise RuntimeError("Trying to initialize a group twice.")
+
+    assert (world_size > 0)
+    assert (rank >= 0)
+    assert (rank < world_size)
+    _group_mgr.create_collective_group(backend, world_size, rank, group_name)
+
+
+def destroy_collective_group(group_name: str = "default") -> None:
+    """Destroy a collective group given its group name."""
+    global _group_mgr
+    _group_mgr.destroy_collective_group(group_name)
+
+
+def get_rank(group_name: str = "default") -> int:
+    """
+    Return the rank of this process in the given group.
+
+    Args:
+        group_name (str): the name of the group to query
+
+    Returns:
+        the rank of this process in the named group,
+        -1 if the group does not exist or the process does
+        not belong to the group.
+    """
+    if not is_group_initialized(group_name):
+        return -1
+    g = _group_mgr.get_group_by_name(group_name)
+    return g.rank
+
+
+def get_world_size(group_name="default") -> int:
+    """
+    Return the size of the collective gropu with the given name.
+
+    Args:
+        group_name: the name of the group to query
+
+    Returns:
+        The world size of the collective group，
+        -1 if the group does not exist or the process does
+        not belong to the group.
+    """
+    if not is_group_initialized(group_name):
+        return -1
+    g = _group_mgr.get_group_by_name(group_name)
+    return g.world_size
+
+
+def allreduce(tensor, group_name: str, op=types.ReduceOp.SUM):
+    """
+    Collective allreduce the tensor across the group with name group_name.
+
+    Args:
+        tensor:
+        group_name (str):
+        op:
+
+    Returns:
+        None
+    """
+    _check_single_tensor_input(tensor)
+    g = _check_and_get_group(group_name)
+    opts = types.AllReduceOptions
+    opts.reduceOp = op
```

Comment:
doesn't this set the global variable?

can we instead create an instance?

Reply:
This is an exposed user API: it does not write; It only reads from the global variable `_group_mgr`.

Reply:
doesn't `types.AllReduceOptions` refer to a global setting?

anyways, i think this is a nit :)